sentence_transformers/trainer.py

```diff
@@ -0,0 +1,299 @@
+            self.dataset_name = metric_key_prefix[5:]
+
+        output = super().evaluation_loop(
+            dataloader=dataloader,
+            description=description,
+            prediction_loss_only=prediction_loss_only,
+            ignore_keys=ignore_keys,
+            metric_key_prefix=metric_key_prefix,
+        )
+
+        # If the evaluator is not defined, we can just return the output
+        if self.evaluator is None:
+            return output
+
+        # If we are training and eval_dataset is a DatasetDict, then we should
+        # 1) only run the evaluator for the first dataset
+        # 2) prefix that only run as "eval", rather than e.g. "eval_multi_nli"
+        if self.is_in_train and isinstance(self.eval_dataset, dict) and metric_key_prefix.startswith("eval_"):
+            if metric_key_prefix[5:] == list(self.eval_dataset.keys())[0]:
+                metric_key_prefix = "eval"
+            else:
+                return output
+
+        evaluator_metrics = self.evaluator(self.model)
+        if not isinstance(evaluator_metrics, dict):
+            evaluator_metrics = {"evaluator": evaluator_metrics}
+
+        # Prefix all keys with metric_key_prefix + '_'
+        for key in list(evaluator_metrics.keys()):
+            if not key.startswith(f"{metric_key_prefix}_"):
+                evaluator_metrics[f"{metric_key_prefix}_{key}"] = evaluator_metrics.pop(key)
+
+        output.metrics.update(evaluator_metrics)
+
+        return output
+
+    def get_train_dataloader(self) -> DataLoader:
+        """
+        Returns the training [`~torch.utils.data.DataLoader`].
+
+        Will use no sampler if `train_dataset` does not implement `__len__`, a random sampler (adapted to distributed
+        training if necessary) otherwise.
+
+        Subclass and override this method if you want to inject some custom behavior.
+        """
+        if self.train_dataset is None:
+            raise ValueError("Trainer: training requires a train_dataset.")
+
+        train_dataset = self.train_dataset
+        data_collator = self.data_collator
+
+        if isinstance(train_dataset, DatasetDict):
+            sizes = [len(ds) for ds in train_dataset.values()]
+            train_dataset = ConcatDataset(train_dataset.values())
+        else:
+            sizes = [len(train_dataset)]
+
+        batch_sampler_class = RoundRobinBatchSampler if self.args.round_robin_sampler else ProportionalBatchSampler
+        dataloader_params = {
+            "collate_fn": data_collator,
+            "num_workers": self.args.dataloader_num_workers,
+            "pin_memory": self.args.dataloader_pin_memory,
+            "persistent_workers": self.args.dataloader_persistent_workers,
+            "batch_sampler": batch_sampler_class(
+                lengths=sizes,
+                batch_size=self.args.train_batch_size,
+                drop_last=self.args.dataloader_drop_last,
+                seed=self.args.seed,
+                trainer=self,
```

Comment:
is there a better way to handle this? currenly we have dataset -> dataloader -> sampler -> trainer -> trainer state (dataset name) is modified on-the-fly in sampler to get the current `dataset_name`, how about in the `BatchSampler`  -> `__iter__`, return a tuple, of the data and dataset_name, and decouple trainer instance from the batch sampler?

Reply:
> how about in the BatchSampler -> __iter__, return a tuple, of the data and dataset_name, and decouple trainer instance from the batch sampler?

I considered this approach, because I was 1) really not a fan of the hacky "batch sampler updates the trainer" approach and 2) it doesn't work on distributed training.
The problem with it is that a `BatchSampler` must always return exactly a list of indices. Otherwise I would also have to override the `DataLoader` so it can understand our updated `BatchSampler`. I decided against this, and instead went with a new solution in bfa29a82539844fa9d3c8d6c603657ab1662cd7f:

Whenever we do distributed training **and** we have multiple training/evaluation datasets, then add a column to each dataset with the dataset name. This requires very slightly more storage (but insignificant compared to the big `input_ids` and `attention_mask` vectors for each sample), but heavily simplifies and reduces the hackiness of the implementation. This also allows for much smoother support for distributed training.

In short, this should work more "normal" now.